core/src/main/java/hivemall/tools/map/MapRouletteUDF.java

```diff
@@ -0,0 +1,192 @@
+        + "Number is a probability value or a positive weight")
+@UDFType(deterministic = false, stateful = false) // it is false because it return value base on probability
+public class MapRouletteUDF extends GenericUDF {
+
+    /**
+     * The map passed in saved all the value and its weight
+     *
+     * @param m A map contains a lot of item as key, with their weight as value
+     * @return The key that computer selected according to key's weight
+     */
+    private Object algorithm(Map<Object, Double> m) {
+        // normalize the weight
+        double sum = 0;
+        for (Map.Entry<Object, Double> entry : m.entrySet()) {
+            sum += entry.getValue();
+        }
+        for (Map.Entry<Object, Double> entry : m.entrySet()) {
+            entry.setValue(entry.getValue() / sum);
+        }
+
+        // sort and generate a number axis
+        List<Map.Entry<Object, Double>> entryList = new ArrayList<>(m.entrySet());
+        Collections.sort(entryList, new MapRouletteUDF.KvComparator());
+        double tmp = 0;
+        for (Map.Entry<Object, Double> entry : entryList) {
+            tmp += entry.getValue();
+            entry.setValue(tmp);
+        }
+
+        // judge last value
+        if (entryList.get(entryList.size() - 1).getValue() > 1.0) {
+            entryList.get(entryList.size() - 1).setValue(1.0);
+        }
+
+        // pick a Object base on its weight
+        double cursor = Math.random();
+        for (Map.Entry<Object, Double> entry : entryList) {
+            if (cursor < entry.getValue()) {
+                return entry.getKey();
+            }
+        }
+        return null;
+    }
+
+    private transient MapObjectInspector mapOI;
+    private transient PrimitiveObjectInspector valueOI;
+
+    @Override
+    public ObjectInspector initialize(ObjectInspector[] arguments) throws UDFArgumentException {
+        if (arguments.length != 1)
+            throw new UDFArgumentLengthException(
+                "Expected one arguments for map_find_max_prob: " + arguments.length);
+        if (arguments[0].getCategory() != ObjectInspector.Category.MAP) {
+            throw new UDFArgumentTypeException(0,
+                "Only map type arguments are accepted for the key but " + arguments[0].getTypeName()
+                        + " was passed as parameter 1.");
+        }
+        mapOI = HiveUtils.asMapOI(arguments[0]);
+        ObjectInspector keyOI = mapOI.getMapKeyObjectInspector();
+
+        //judge valueOI is a number
+        valueOI = (PrimitiveObjectInspector) mapOI.getMapValueObjectInspector();
+        switch (valueOI.getTypeName()) {
+            case INT_TYPE_NAME:
+            case DOUBLE_TYPE_NAME:
+            case BIGINT_TYPE_NAME:
+            case FLOAT_TYPE_NAME:
+            case SMALLINT_TYPE_NAME:
+            case TINYINT_TYPE_NAME:
+            case DECIMAL_TYPE_NAME:
+            case STRING_TYPE_NAME:
```

Comment:
I think `STRING_TYPE_NAME` can be removed because it's better to assume numbers as values in map_roulette.

Reply:
But when pass an `empty map` or a map full of ` {null, null}` will get string type there. To be honest, I don't know why, just add a `STRING_TYPE_NAME` there.